### ISSUE TITLE
Network latency increases with joins

### ISSUE THREAD

**Issue by chrbala** [prisma/prisma#12582]
### Bug description

Per https://github.com/prisma/prisma/issues/5043#issuecomment-1081722399, I am making this issue. I have found that joins have worse performance as the client <-> database latency increases. 

### How to reproduce

Check out the [typescript/testing-express example in my prisma-examples fork](https://github.com/chrbala/prisma-examples/tree/latency-repro/typescript/testing-express)

##### Initial setup
`yarn install`

##### Testing with a local database
Add a .env file that points to a local mysql database. e.g. `DB_URL=mysql://...`
`yarn run seed`
`yarn start`
`curl http://localhost:3000/test`

Observe the shallow and deep query times in the server logs.

##### Testing with a remote database
Add a .env file that points to a remote mysql database, such as one on PlanetScale. e.g. `DB_URL=mysql://...`
`yarn run seed`
`yarn start`
`curl http://localhost:3000/test`

Observe the logs as above.

##### Notes
* It may also be possible to reproduce this issue by means of injecting latency into the connection with means other than a distant server. Anything that makes request times take longer should work.
* In each case above, the first query made via `curl` will be inaccurate because the database must establish the connection.

### Expected behavior

We should expect that using joins does not increase query latency by integer multiples. Instead, we see that each join included will increase the latency by roughly the cost of 1 round trip.

#### Results
The below is the results from 5 sample requests to the test server. The first request was removed from each set because it includes database connection time. The local queries are very fast at <5ms, but the remote ones get slow. Specifically, the deep queries take ~10x the time as the shallow queries. This probably happens locally too, but the connection is so fast that the difference is not perceptible.

The ~10x time increase for the deep queries is not a coincidence. We request 10 models in a waterfall, so it stands to reason that the request time is 10x as long. The query time is roughly: [local overhead] + [join depth] x [network overhead]. If the queries could be executed all on the SQL server, then query time would be closer to [local overhead] + [network overhead]. Specifically for the data in this repro, making the queries on the server would reduce the remote deep queries from ~1000ms to roughly the same amount of time as the shallow remote queries, or ~100ms.

##### Local
shallow 1
deep 3
shallow 1
deep 2
shallow 1
deep 2
shallow 1
deep 2
shallow 1
deep 2

##### Remote
In this case, I am on the US West coast making requests to PlanetScale on us-east-1.

shallow 105
deep 1040
shallow 109
deep 1043
shallow 107
deep 1212
shallow 215
deep 1753
shallow 103
deep 1068

### Prisma information

# Schema
```prisma
datasource db {
  provider             = "mysql"
  url                  = env("DB_URL")
  referentialIntegrity = "prisma"
}

generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}

model One {
  id    Int     @default(autoincrement()) @id
  next Two[]
}

model Two {
  id    Int     @default(autoincrement()) @id
  previous One @relation(fields: [id], references: [id])
  next Three[]
}

model Three {
  id    Int     @default(autoincrement()) @id
  previous Two @relation(fields: [id], references: [id])
  next Four[]
}

model Four {
  id    Int     @default(autoincrement()) @id
  previous Three @relation(fields: [id], references: [id])
  next Five[]
}

model Five {
  id    Int     @default(autoincrement()) @id
  previous Four @relation(fields: [id], references: [id])
  next Six[]
}

model Six {
  id    Int     @default(autoincrement()) @id
  previous Five @relation(fields: [id], references: [id])
  next Seven[]
}


model Seven {
  id    Int     @default(autoincrement()) @id
  previous Six @relation(fields: [id], references: [id])
  next Eight[]
}

model Eight {
  id    Int     @default(autoincrement()) @id
  previous Seven @relation(fields: [id], references: [id])
  next Nine[]
}

model Nine {
  id    Int     @default(autoincrement()) @id
  previous Eight @relation(fields: [id], references: [id])
  next Ten[]
}

model Ten {
  id    Int     @default(autoincrement()) @id
  previous Nine @relation(fields: [id], references: [id])
  value String
}
```

# Shallow query
```ts
prisma.one.findFirst()
```

# Deep query
```ts
prisma.one.findFirst({
    include: {
      next: {
        include: {
          next: {
            include: {
              next: {
                include: {
                  next: {
                    include: {
                      next: {
                        include: {
                          next: {
                            include: {
                              next: {
                                include: {
                                  next: {
                                    include: {
                                      next: true,
                                    },
                                  },
                                },
                              },
                            },
                          },
                        },
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    },
  })
```

### Environment & setup

- OS: Mac OS
- Database: MySQL
- Node.js version: v16.13.0

### Prisma Version

```
3.7.0
```

**Comment by remorses** [comment]
Prisma doesn’t have joins support, when you query relations it makes many serial queries 
#5136

**Comment by Velozity** [comment]
+1 for the love of god, blease

**Comment by janpio** [comment]
Note: Probably not related to `referentialIntegrity=prisma` but just by chance reported for PlanetScale (where this is used).

**Comment by Jolg42** [comment]
Trying the reproduction from the git repo provided on a local machine with a docker image of MySQL shows that a `deep` query is slower but there is a lot a variability

```
On 4.4.0
Running: watch -n 1 curl http://localhost:3000/test

referentialIntegrity = "prisma"
shallow 3
deep 17
shallow 4
deep 10
shallow 6
deep 23
shallow 3
deep 26
shallow 4
deep 12
shallow 3
deep 12
shallow 4
deep 9
shallow 2
deep 10
shallow 2
deep 12
shallow 2
deep 11
shallow 2
deep 10
shallow 2
deep 10

Without referentialIntegrity = "prisma" (= using foreign keys)
shallow 3
deep 22
shallow 5
deep 21
shallow 4
deep 13
shallow 3
deep 16
shallow 2
deep 11
shallow 2
deep 11
shallow 2
deep 11
shallow 3
deep 10
shallow 2
deep 11
shallow 2
deep 12
shallow 2
deep 12
shallow 19
deep 14
shallow 3
deep 28
shallow 4
deep 29
shallow 6
deep 17
```

I cannot conclude that the `referentialIntegrity` preview feature as any effect when set to `prisma`.

**Comment by janpio** [comment]
A note: The main query about JOINs usage, where we will post updates and where a lot of discussion is happening right now is https://github.com/prisma/prisma/issues/5184. Make sure to follow that as well, although I will of course also try to update here if something relevant changes. Thanks.

**Comment by Weakky** [comment]
Hey folks,

We’re excited to let you know that we’ve just added support for `JOIN`s to resolve relations in [Prisma 5.7.0](https://github.com/prisma/prisma/releases/tag/5.7.0) for PostgreSQL and CockroachDB, under the `relationJoins` preview feature.

Once you enable that feature, `JOIN`s will be used to resolve one-to-one, one-to-many and many-to-many relations. Most of the queries listed on [this documentation page](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-reads) will essentially now use joins. This means that, when possible, we now resolve your query in a single round-trip as opposed to the N round-trips we used to make for each relation.

I say *when possible*, because we fallback to the old query-based mechanism when you:

- Use `distinct`
- Use `cursor`
- Use `_count` in your selection set

This is the **very first step** to bring `JOIN`s to Prisma for querying relations. We plan on removing the current limitations and adding support for other databases in future releases.

Changing the way Prisma resolves relations is not a trivial change and is likely not going to work 100% on day one, so please bear with us as we continuously improve the feature, both in terms of bugs and performance improvements on the SQL queries we generate.

If you are curious, we have provided a bunch of Prisma query examples together with their associated SQL query output in our [latest release notes](https://github.com/prisma/prisma/releases/tag/5.7.0)

Finally, we have a dedicated [preview feature feedback issue here](https://github.com/prisma/prisma/discussions/22288) if you have any thoughts about it.

Thank you all for your patience!

**Comment by matjaeck** [comment]
The title of this issue is probably misleading because the benchmarks have at that time not used `JOIN`s (or am I missing something?). Now that we have this feature it should be re-evaluated if it's true that "Network latency increases with joins" (hopefully not).

**Comment by janpio** [comment]
I renamed the issue to reflect reality, which is that selecting nested increased the network latency. Should have done that before, thanks for the reminder.

**Comment by chrbala** [comment]
Right, there were never actual joins being tested here. I believe joins are still not being used in this example, however, because this example uses `MySQL` and not `PostgreSQL` or `CockroachDB`. I would really like to see proper joins used for `MySQL` because that would mean proper support for [PlanetScale](https://planetscale.com).

**Comment by aqrln** [comment]
Hey @chrbala and others! As a follow up to the previous comment by @Weakky, in Prisma 5.10.0 we have released support for MySQL 8.0.14 and newer in the `relationJoins` preview feature. MySQL currently still falls back to the old strategy if a query contains `distinct` or `cursor`. Please try it out and let us know if it solves the problem for you and if you are seeing a performance improvement 🙏
